src/compute/src/render/join/linear_join.rs

```diff
@@ -313,3 +316,442 @@ where
+        batch: C2,
+        batch_storage: C2::Storage,
+        capability: Capability<T>,
+    ) -> Self {
+        Deferred {
+            trace,
+            trace_storage,
+            batch,
+            batch_storage,
+            capability,
+            done: false,
+            temp: Vec::new(),
+        }
+    }
+
+    fn work_remains(&self) -> bool {
+        !self.done
+    }
+
+    /// Process keys until at least `fuel` output tuples produced, or the work is exhausted.
+    #[inline(never)]
+    fn work<L, I>(
+        &mut self,
+        output: &mut OutputHandle<T, (D, T, Diff), Tee<T, (D, T, Diff)>>,
+        mut logic: L,
+        fuel: &mut usize,
+    ) where
+        I: IntoIterator<Item = (D, T, Diff)>,
+        L: FnMut(&C1::Key, &C1::Val, &C2::Val, &T, &C1::R, &C2::R) -> I,
+    {
+        let meet = self.capability.time();
+
+        let mut session = output.session(&self.capability);
+
+        let trace_storage = &self.trace_storage;
+        let batch_storage = &self.batch_storage;
+
+        let trace = &mut self.trace;
+        let batch = &mut self.batch;
+
+        let temp = &mut self.temp;
+
+        while batch.key_valid(batch_storage) && trace.key_valid(trace_storage) {
+            match trace.key(trace_storage).cmp(batch.key(batch_storage)) {
+                Ordering::Less => trace.seek_key(trace_storage, batch.key(batch_storage)),
+                Ordering::Greater => batch.seek_key(batch_storage, trace.key(trace_storage)),
+                Ordering::Equal => {
+                    assert_eq!(temp.len(), 0);
+
+                    // Populate `temp` with the results, as long as fuel remains.
+                    let key = batch.key(batch_storage);
+                    while let Some(val1) = trace.get_val(trace_storage) {
+                        while let Some(val2) = batch.get_val(batch_storage) {
+                            trace.map_times(trace_storage, |time1, diff1| {
+                                let time1 = time1.join(meet);
+                                batch.map_times(batch_storage, |time2, diff2| {
+                                    let time = time1.join(time2);
+                                    for (d, t, r) in logic(key, val1, val2, &time, diff1, diff2) {
+                                        temp.push(((d, t), r));
+                                    }
+                                });
+                            });
+                            batch.step_val(batch_storage);
+                        }
+                        batch.rewind_vals(batch_storage);
+                        trace.step_val(trace_storage);
+
+                        // TODO: This consolidation is optional, and it may not be very
+                        //       helpful. We might try harder to understand whether we
+                        //       should do this work here, or downstream at consumers.
+                        // TODO: Perhaps `thinker` should have the buffer, do smarter
```

Comment:
`thinker` no longer exists, maybe reword this comment?

Reply:
I removed the comment. I don't think with the simple nested-loop matching approach there is not much we can do to consolidate smarter.